docs/design-documents/20240812-recover-from-pipeline-errors.md

```diff
@@ -0,0 +1,194 @@
+
+## Implementation plan
+
+We will tackle this feature in separate phases, to make sure that the most
+valuable parts are delivered right away, while the non-critical improvements are
+added later.
+
+### Phase 1
+
+In this phase we will deliver the most critical functionality,
+the [recovery mechanism](#recovery-mechanism)
+and [pipeline state management](#pipeline-state-management). This is the minimum
+we need to implement to provide value to our users, while only changing the
+Conduit internals.
+
+This phase includes the following tasks:
+
+- Implement recovery mechanism as described in the topic above including the
+  backoff retry.
+- When a pipeline is restarting it should be in the `recovering` state
+- Add the introduced `recovering` state to the API definition
+- Make the backoff retry default parameters configurable _globally_ (in
+  `conduit.yml`)
+
+### Phase 2
+
+This phase includes changes in the connector SDK and gives the connectors the
+ability to circumvent the recovery mechanism when it makes sense to do so
+using [error classification](#error-classification).
+
+The tasks in this phase are:
+
+- Change the connector SDK to format error messages as JSON and include an error
+  code in the error. Document the behavior.
+- Conduit should parse errors coming from connectors as JSON. If they are valid
+  JSON and include the error code field, that should be used to determine if an
+  error is fatal or not.
+  - Backwards compatibility with older connectors must be maintained, so plain
+    text errors must still be accepted.
+- Define error codes (or ranges, like HTTP errors) and define which ranges are
+  transient or fatal.
+- Add utilities in the connector SDK to create fatal or transient errors.
+  - This should work even if errors are wrapped before being returned.
+- On a fatal error, Conduit should skip the recovery mechanism and move the
+  pipeline state into `degraded` right away.
+
+### Phase 3
+
+In this phase we will sand off any remaining sharp edges and add more visibility
+and control to the user. This
+includes [record deduplication](#record-deduplication) and
+the [audit log](#audit-log), which are bigger topics and will require a separate
+design document.
+
+## Open questions
+
+- **Should failed processors trigger a retry, or should the pipeline be
+  considered failed immediately?**
+  - I am leaning towards making processor faults fatal by default, as most of
+    the processors are stateless and always behave the same. Processors, which
+    connect to a 3rd party resource and can experience transient errors, should
+    implement retrials internally (like the `webhook.http` processor). Once we
+    implement [error classification](#error-classification), processors should
+    be allowed to return transient errors which should be treated as such.
+- **How does this functionality interact with a DLQ, especially with a nack
+  threshold?**
```

Comment:
To me, this is the biggest question. How to pair the pipeline recovery feature with a DLQ. Considering that I value more providing good defaults on pipeline recovery, keeping the default of a nack threshold of 0 on DLQs makes perfect sense. We need to be careful of documenting well where this pipeline recovery takes effect so it's clear to the user that there are two different levels of tolerance. One for records that can't be processed successfully, and another for once a pipeline errors in a more _expected_ way.

Reply:
Update: We discussed this the other day at standup at length. I don't think we came up with a great conclusion, so we'll need to continue thinking about this one while in development.